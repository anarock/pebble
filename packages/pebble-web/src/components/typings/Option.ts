import { ControlProps } from "./Control";
import { Omit } from "utility-types";
import * as React from "react";
import { Interpolation } from "./emotionCustom";

export interface OptionProps<OptionType>
  extends Omit<ControlProps<OptionType>, "checked" | "type"> {
  isActive?: boolean;
  isSelected?: boolean;
  multiSelect?: boolean;
  labelStyles?: Interpolation;
  rightElement?: (props: OptionProps<OptionType>) => React.ReactNode;
}
