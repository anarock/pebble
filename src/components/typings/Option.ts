import { ControlProps } from "./Control";
import { Omit } from "utility-types";
import * as React from "react";

export interface OptionProps<OptionType>
  extends Omit<ControlProps<OptionType>, "checked" | "type"> {
  isActive?: boolean;
  isSelected?: boolean;
  multiSelect?: boolean;
  labelClassName?: string;
  rightElement?: (props: OptionProps<OptionType>) => React.ReactNode;
}
