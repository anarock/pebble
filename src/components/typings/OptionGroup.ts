import * as React from "react";
import { SearchProps } from "./Search";
import { Omit } from "utility-types";

export interface OptionGroupProps {
  className?: string;
  searchBox?: boolean;
  searchBoxProps?: Omit<SearchProps, "type">;
  isSelected: (value: React.ReactText) => boolean;
  handleChange: (
    args: { value: React.ReactText; checked: boolean },
    e: React.SyntheticEvent | Event
  ) => void;
  multiSelect?: boolean;
}

export type CommonProps = Omit<
  OptionGroupProps,
  "multiSelect" | "isSelected" | "handleChange"
>;

export interface Extras {
  props: CommonProps;
  event: React.SyntheticEvent | Event;
}

export interface OptionGroupState {
  selected: number;
  isScrolled: boolean;
}
