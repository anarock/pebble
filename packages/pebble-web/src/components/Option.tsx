import * as React from "react";
import { OptionProps } from "./typings/Option";
import Control from "./shared/Control";
import {
  activeRow,
  rowWrapper,
  selectedRow,
  labelWrap
} from "./styles/Options.styles";
import Ink from "react-ink";
import { colors } from "pebble-shared";

const defaultProps = {
  rightElement: ({
    isSelected,
    multiSelect
  }: {
    isSelected: boolean;
    multiSelect: boolean;
  }) => {
    const iconClass = [
      "pi",
      isSelected && "pi-checkbox-selected",
      !isSelected && "pi-checkbox-unselected"
    ]
      .filter(Boolean)
      .join(" ");
    return multiSelect ? (
      <i
        className={iconClass}
        css={{
          marginLeft: "10px",
          color: isSelected ? colors.violet.base : colors.gray.light,
          fontSize: "20px"
        }}
      />
    ) : null;
  }
};

class Option<OptionType> extends React.Component<
  OptionProps<OptionType> & Required<typeof defaultProps>
> {
  static defaultProps = defaultProps;
  render() {
    const {
      label,
      isActive,
      isSelected,
      rightElement,
      labelStyles,
      styles
    } = this.props;
    const _class = [
      rowWrapper,
      !!isActive && activeRow,
      !!isSelected && selectedRow,
      styles
    ];
    return (
      <Control
        {...this.props}
        checked={this.props.isSelected}
        type={this.props.multiSelect ? "checkbox" : "radio"}
        styles={_class}
      >
        {() => {
          return (
            <>
              <div css={[labelWrap, labelStyles]}>{label}</div>
              {rightElement(this.props)}
              <Ink />
            </>
          );
        }}
      </Control>
    );
  }
}

export default Option;
