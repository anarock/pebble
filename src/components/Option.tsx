import * as React from "react";
import { OptionProps } from "./typings/Option";
import Control from "./shared/Control";
import { cx } from "emotion";
import { activeRow, rowWrapper, selectedRow } from "./styles/Options.styles";
import Ink from "react-ink";
import { colors, mixins } from "../theme";
import { Component } from "react";

class Option extends Component<OptionProps> {
  static defaultProps: Partial<OptionProps> = {
    rightElement: ({ isSelected, multiSelect }) => {
      const iconClass = cx({
        "icon-checkbox-selected": isSelected,
        "icon-checkbox-unselected": !isSelected
      });
      return multiSelect ? (
        <i
          style={{
            color: isSelected ? colors.violet.base : colors.gray.light
          }}
          className={iconClass}
        />
      ) : null;
    }
  };

  render() {
    return (
      <Control
        {...this.props}
        checked={this.props.isSelected}
        type={this.props.multiSelect ? "checkbox" : "radio"}
      >
        {() => {
          const { label, isActive, isSelected, rightElement } = this.props;

          const _class = cx(rowWrapper, {
            [activeRow]: isActive,
            [selectedRow]: isSelected
          });

          return (
            <div className={_class} style={mixins.flexSpaceBetween}>
              {label} {rightElement(this.props)}
              <Ink />
            </div>
          );
        }}
      </Control>
    );
  }
}

export default Option;
