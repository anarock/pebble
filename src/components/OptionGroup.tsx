import * as React from "react";
import { optionsWrapper } from "./styles/Options.styles";
import { OptionProps } from "./typings/Option";
import { getSelectedCheckboxes } from "./utils/getSelectedCheckboxes";
import { findDOMNode } from "react-dom";
import { OptionGroupProps, OptionGroupState } from "./typings/OptionGroup";

export default class OptionGroup extends React.PureComponent<
  OptionGroupProps,
  OptionGroupState
> {
  optionRef: React.RefObject<HTMLDivElement> = React.createRef();

  state = {
    selected: 0
  };

  private handleKeyPress = (e: KeyboardEvent) => {
    const { children } = this.props;
    const { selected } = this.state;

    if (e.which === 13) {
      // @ts-ignore
      this.handleChange(children[selected].props);
    }

    this.setState(
      () => {
        let _selected = selected;
        if (e.which === 40) {
          e.preventDefault();
          _selected = Math.min(
            _selected + 1,
            React.Children.count(children) - 1
          );
        }
        if (e.which === 38) {
          e.preventDefault();
          _selected = Math.max(_selected - 1, 0);
        }

        return { selected: _selected };
      },
      () => {
        if (this.optionRef.current && (e.which === 40 || e.which === 38)) {
          findDOMNode(this[`option-ref-${selected}`].current).scrollIntoView({
            behavior: "smooth"
          });
        }
      }
    );
  };

  componentDidMount() {
    document.addEventListener("keydown", this.handleKeyPress);
  }

  componentWillUnmount() {
    document.removeEventListener("keydown", this.handleKeyPress);
  }

  handleChange = ({ value, checked }) => {
    const { multi, onChange, selected } = this.props;

    if (multi) {
      // @ts-ignore
      onChange(getSelectedCheckboxes(value, selected), this.props);
    } else {
      onChange(checked ? value : undefined, this.props);
    }
  };

  private isSelected = value => {
    const { multi, selected } = this.props;
    // @ts-ignore
    return multi ? selected.indexOf(value) >= 0 : selected === value;
  };

  render() {
    const _children = React.Children.map(
      this.props.children,
      (option: React.ReactElement<OptionProps>, i) => {
        this[`option-ref-${i}`] = React.createRef();
        return React.cloneElement(option, {
          onChange: this.handleChange,
          isActive: this.state.selected === i,
          isSelected: this.isSelected(option.props.value),
          multi: this.props.multi,
          // @ts-ignore
          ref: this[`option-ref-${i}`]
        });
      }
    );

    return (
      <div ref={this.optionRef} className={optionsWrapper}>
        {_children}
      </div>
    );
  }
}
