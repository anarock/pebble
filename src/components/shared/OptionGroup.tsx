import * as React from "react";
import * as ReactDOM from "react-dom";
import { OptionProps } from "../typings/Option";
import { OptionGroupProps, OptionGroupState } from "../typings/OptionGroup";
import scrollIntoView from "scroll-into-view-if-needed";
import { cx } from "emotion";
import Search from "../Search";
import {
  searchBoxScrolledStyle,
  searchBoxWrapper,
  optionsWrapper,
  searchBoxHeight
} from "../styles/OptionGroup.styles";

class OptionGroup<OptionType> extends React.PureComponent<
  OptionGroupProps<OptionType>,
  OptionGroupState
> {
  optionRef: React.RefObject<HTMLDivElement> = React.createRef();
  optionsRefsSet = new Map<number, React.RefObject<React.ReactInstance>>();
  observer?: IntersectionObserver;

  state: Readonly<OptionGroupState> = {
    highlighted: -1,
    isScrolled: false
  };

  private handleKeyPress = (e: React.KeyboardEvent<HTMLInputElement>) => {
    const { handleChange, isSelected } = this.props;
    const children = React.Children.toArray(this.props.children);
    const { highlighted } = this.state;
    const { which } = e;

    if (which === 13 && children && children[highlighted]) {
      // Enter key
      // @ts-ignore
      const { value } =
        // @ts-ignore
        children && children[highlighted] && children[highlighted].props;

      handleChange(
        {
          value,
          checked: !isSelected(value)
        },
        e
      );
    }

    this.setState(
      () => {
        let _highlighted = highlighted;
        if (which === 40) {
          _highlighted = Math.min(
            _highlighted + 1,
            React.Children.count(children) - 1
          );
        }
        if (which === 38) {
          _highlighted = Math.max(_highlighted - 1, 0);
        }

        return { highlighted: _highlighted };
      },
      () => {
        const currentRef = this.optionsRefsSet.get(highlighted);
        if (
          this.optionRef.current &&
          (which === 40 || which === 38) &&
          currentRef &&
          currentRef.current
        ) {
          const element = ReactDOM.findDOMNode(currentRef.current) as Element;
          if (element) {
            scrollIntoView(element, {
              behavior: "smooth",
              boundary: this.optionRef.current
            });
          }
        }
      }
    );
  };

  componentDidMount() {
    this.observer = new IntersectionObserver(
      entries => {
        this.setState({
          isScrolled: entries[0].intersectionRatio < 0.9
        });
      },
      {
        root: this.optionRef.current,
        threshold: 0.9
      }
    );

    if (
      this.optionRef.current &&
      this.optionRef.current.childNodes &&
      this.optionRef.current.childNodes.length
    ) {
      this.observer.observe(this.optionRef.current.childNodes[0] as Element);
    }
  }

  componentWillUnmount() {
    if (this.observer) {
      this.observer.disconnect();
    }
  }

  render() {
    const {
      searchBox,
      children,
      multiSelect,
      className,
      isSelected,
      handleChange,
      searchBoxProps
    } = this.props;
    const { isScrolled, highlighted } = this.state;

    const _children = React.Children.map(children, (_option, i) => {
      // `_option as React.ReactElement<OptionProps>` is a hack
      // Because React does not allow us to specify what sort of elements
      // you can allow as children and leaves it on you to figure out
      // all various types of children provided.
      const option = _option as React.ReactElement<OptionProps<OptionType>>;
      let ref = this.optionsRefsSet.get(i);
      if (!ref) {
        ref = React.createRef<HTMLDivElement>();
        this.optionsRefsSet.set(i, ref);
      }
      return React.cloneElement(option, {
        onChange: handleChange,
        isActive: highlighted === i,
        isSelected: isSelected(option.props.value),
        multiSelect,
        // @ts-ignore
        ref
      });
    });

    const searchBoxClassName = cx(searchBoxWrapper, {
      [searchBoxScrolledStyle]: isScrolled
    });

    return (
      <React.Fragment>
        {searchBox && searchBoxProps && (
          <div className={searchBoxClassName}>
            <Search
              type="small"
              {...searchBoxProps}
              inputProps={{
                ...(searchBoxProps && searchBoxProps.inputProps),
                onKeyDown: this.handleKeyPress,
                autoFocus: true
              }}
            />
          </div>
        )}
        {!!React.Children.count(children) && (
          <div
            ref={this.optionRef}
            style={{
              paddingTop: searchBox ? searchBoxHeight : undefined
            }}
            className={cx(optionsWrapper, className)}
            role={multiSelect ? "group" : "radiogroup"}
            data-test-id="optiongroup"
          >
            {_children}
          </div>
        )}
      </React.Fragment>
    );
  }
}

export default OptionGroup;
