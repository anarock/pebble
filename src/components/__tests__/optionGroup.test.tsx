import * as React from "react";
import renderer from "react-test-renderer";
import { mount } from "enzyme";
import sinon from "sinon";

import Option from "../Option";
import OptionGroup from "../shared/OptionGroup";

function getComponent(props: Partial<OptionGroup["props"]> = {}) {
  return (
    <OptionGroup isSelected={() => false} handleChange={() => {}} {...props}>
      {[...Array(12)].map((_, i) => (
        <Option value={`option-${i}`} label={`Option ${i}`} key={i} />
      ))}
    </OptionGroup>
  );
}

describe("Component: OptionGroup", () => {
  test("default snapshot", () => {
    const component = renderer.create(getComponent());
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test("with searchbox", () => {
    const component = renderer.create(
      getComponent({
        searchBox: true,
        searchBoxProps: {
          value: "",
          placeholder: "Search",
          onChange: () => {}
        }
      })
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test("multi option group", () => {
    const component = renderer.create(
      getComponent({
        multiSelect: true,
        searchBox: true,
        searchBoxProps: {
          value: "",
          placeholder: "Search",
          onChange: () => {}
        }
      })
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test("click events", () => {
    const spy = sinon.spy();
    const component = mount(
      getComponent({
        multiSelect: true,
        handleChange: spy
      })
    );

    component
      .find(Option)
      .at(0)
      .simulate("click");

    expect(spy.calledOnce).toBeTruthy();

    expect(
      spy.calledWith({
        value: "option-0",
        checked: true
      })
    ).toBeTruthy();
  });

  test("keyboard events", () => {
    const spy = sinon.spy();
    const component = mount(
      getComponent({
        multiSelect: true,
        handleChange: spy,
        searchBox: true,
        searchBoxProps: {
          value: "",
          placeholder: "Search",
          onChange: () => {}
        }
      })
    );

    expect(spy.notCalled).toBeTruthy();

    const input = component.find("input");

    input
      .simulate("focus")
      .simulate("keydown", {
        which: 40 // down
      })
      .simulate("keydown", {
        which: 40 // down
      })
      .simulate("keydown", {
        which: 13 // enter
      });

    expect(
      spy.calledWith({
        value: "option-1",
        checked: true
      })
    ).toBeTruthy();

    input
      .simulate("keydown", {
        which: 40 // down
      })
      .simulate("keydown", {
        which: 40 // down
      })
      .simulate("keydown", {
        which: 38 // down
      })
      .simulate("keydown", {
        which: 13 // enter
      });

    expect(
      spy.calledWith({
        value: "option-2",
        checked: true
      })
    ).toBeTruthy();
  });
});
