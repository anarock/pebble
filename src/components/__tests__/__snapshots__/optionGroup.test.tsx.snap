// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OptionGroup Component snapshot in selected state 1`] = `
.emotion-6 {
  background-color: #ffffff;
  border-radius: 3px;
  position: relative;
  width: inherit;
  z-index: 9;
  max-height: 352px;
  min-width: 200px;
  overflow-y: auto;
  padding: 20px 1px;
}

.emotion-1 {
  outline: none;
  position: relative;
}

.emotion-2 {
  font-weight: 400;
  color: #101721;
  font-size: 14px;
  line-height: 10px;
  cursor: pointer;
  padding: 0 18px;
  position: relative;
  z-index: 999;
  overflow-x: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-content: initial;
  -ms-flex-line-pack: initial;
  align-content: initial;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: 52px;
}

.emotion-2:last-of-type {
  border: 0;
}

.emotion-2:hover {
  background-color: #F7F7F9;
}

.emotion-0 {
  font-weight: 400;
  color: #101721;
  font-size: 14px;
  line-height: 10px;
  cursor: pointer;
  padding: 0 18px;
  position: relative;
  z-index: 999;
  overflow-x: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-content: initial;
  -ms-flex-line-pack: initial;
  align-content: initial;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: 52px;
  background-color: #F7F7F9;
  color: #101721;
  font-weight: 500;
  font-size: 14px;
}

.emotion-0:last-of-type {
  border: 0;
}

.emotion-0:hover {
  background-color: #F7F7F9;
}

<div
  className="emotion-6"
  style={
    Object {
      "paddingTop": undefined,
    }
  }
>
  <div
    aria-checked={true}
    className="emotion-1"
    onClick={[Function]}
    role="radio"
    tabIndex={0}
  >
    <div
      className="emotion-0"
      style={
        Object {
          "alignContent": "initial",
          "display": "flex",
          "flexDirection": "row",
          "justifyContent": "space-between",
        }
      }
    >
      I am an option
       
      <canvas
        height={0}
        onDragOver={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        style={
          Object {
            "borderRadius": "inherit",
            "height": "100%",
            "left": 0,
            "position": "absolute",
            "top": 0,
            "width": "100%",
          }
        }
        width={0}
      />
    </div>
  </div>
  <div
    aria-checked={false}
    className="emotion-1"
    onClick={[Function]}
    role="radio"
    tabIndex={-1}
  >
    <div
      className="emotion-2"
      style={
        Object {
          "alignContent": "initial",
          "display": "flex",
          "flexDirection": "row",
          "justifyContent": "space-between",
        }
      }
    >
      I am an option
       
      <canvas
        height={0}
        onDragOver={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        style={
          Object {
            "borderRadius": "inherit",
            "height": "100%",
            "left": 0,
            "position": "absolute",
            "top": 0,
            "width": "100%",
          }
        }
        width={0}
      />
    </div>
  </div>
  <div
    aria-checked={false}
    className="emotion-1"
    onClick={[Function]}
    role="radio"
    tabIndex={-1}
  >
    <div
      className="emotion-2"
      style={
        Object {
          "alignContent": "initial",
          "display": "flex",
          "flexDirection": "row",
          "justifyContent": "space-between",
        }
      }
    >
      I am an option
       
      <canvas
        height={0}
        onDragOver={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        style={
          Object {
            "borderRadius": "inherit",
            "height": "100%",
            "left": 0,
            "position": "absolute",
            "top": 0,
            "width": "100%",
          }
        }
        width={0}
      />
    </div>
  </div>
</div>
`;
